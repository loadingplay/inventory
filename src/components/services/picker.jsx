class PickerStore {
  constructor() {
    if(!PickerStore.instance) {
      let storage_data;

      try {
        storage_data = JSON.parse(localStorage.getItem('picker-data'));
      } catch(ex) {
        console.log(ex);
      }

      this.data = storage_data ? storage_data:{};
      PickerStore.instance = this;
    }

    return PickerStore.instance;
  }

  set = (sku, value) => {
    this.data[sku] = value;
    localStorage.setItem('picker-data', JSON.stringify(this.data));
  }

  get = (sku) => {
    let number = parseInt(this.data[sku]);
    if (isNaN(number)) return 0;
    return number;
  }

}

const instance = new PickerStore();
Object.freeze(instance);

export default instance;
