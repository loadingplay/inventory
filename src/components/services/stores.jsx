export default class StoreLoader {
  constructor(access_token) {
    this.access_token = access_token;
  }

  loadStoreList = async (page, cellar_id) => {
    let response, json_data;

    response = await fetch(
      `https://replenishments.loadingplay.com/replenishment?items=100&page=${page}&cellar_id=${cellar_id}`,
      {
        "headers": {
          "Authorization": `Bearer ${this.access_token}`
        }
      });

    json_data = await response.json();
    return json_data;
  }
}
