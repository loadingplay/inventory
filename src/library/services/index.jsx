export * from "./picker";
export * from "./stock";
export * from "./stores";
