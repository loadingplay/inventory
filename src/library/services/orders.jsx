
export class Orders {

  static API_URL = "https://apibodegas.loadingplay.com";

  constructor(access_token) {
    this.access_token = access_token;
  }

  create = async (post_data) => {
    let response,
      json_data;

    response = await fetch(
      Orders.API_URL + "/v1/order?",
      {
        method: "POST",
        headers: {
          "Authorization": "Bearer " + this.access_token,
          'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: this._convertToQueryString(post_data)
      }
    );

    json_data = await response.json();
    return json_data
  }

  update = async (order_id, put_data) => {
    let response,
      json_data;

    response = await fetch(
      Orders.API_URL + "/v1/order/" + order_id,
      {
        method: "PUT",
        headers: {
          "Authorization": "Bearer " + this.access_token,
          'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: this._convertToQueryString(put_data)
      }
    );

    json_data = await response.json();
    return json_data;
  }

  get = async (order_id) => {
    let response,
      json_data;

    response = await fetch(
      Orders.API_URL + '/v1/order/' + order_id,
      {
        method: "GET",
        headers: {
          "Authorization": `Bearer ${this.access_token}`
        }
      }
    );

    json_data = await response.json();
    return json_data
  }

  _convertToQueryString = (json_data) => {
    return Object.keys(json_data).map(key => key + '=' + json_data[key]).join('&')
  }
}
