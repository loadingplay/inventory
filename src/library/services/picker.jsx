class _PickerStore {
  constructor() {
    if(!_PickerStore.instance) {
      let storage_data;

      try {
        if (typeof window !== 'undefined' && window.localStorage)
          storage_data = JSON.parse(localStorage.getItem('picker-data'));
      } catch(ex) {
        console.log(ex);
      }

      this.data = storage_data ? storage_data:{};
      _PickerStore.instance = this;
    }

    return _PickerStore.instance;
  }

  set = (cellar_id, sku, value, hq_inventory, current_inventory, suggested) => {
    if (!this.data || !(cellar_id in this.data) || this.data.cellar_id === "") {
      this.data[cellar_id] = {};
    }
    if (value === 0)
      delete this.data[cellar_id][sku];
    else
      this.data[cellar_id][sku] = {
        value,
        hq_inventory,
        current_inventory,
        suggested
      };

    if (typeof window !== 'undefined' && window.localStorage)
      localStorage.setItem('picker-data', JSON.stringify(this.data));
  }

  get = (cellar_id, sku) => {
    if (!this.data || !(cellar_id in this.data)) return 0;

    if (sku === undefined) return this.data[cellar_id];
    if (this.data[cellar_id][sku] === undefined) return 0;

    let number = parseInt(this.data[cellar_id][sku].value);
    if (isNaN(number)) return 0;

    return number;
  }

  clear = (cellar_id) => {
    if (!this.data || !(cellar_id in this.data)) return;
    this.data[cellar_id] = undefined;
    delete this.data[cellar_id];
    if (typeof window !== 'undefined' && window.localStorage)
      localStorage.setItem('picker-data', JSON.stringify(this.data));
  }

  getAll = () => {
    return this.data;
  }

}

export const PickerStore = new _PickerStore();
