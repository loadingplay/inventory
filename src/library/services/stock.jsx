var Loader = function (access_token) {
  this.access_token = access_token;
  this.instances = 0;
  this.max_instances = 15;
  this.chunk_size = 10;
};

Loader.prototype.load = function (cellar_id_list, sku_list) {
  var loader = this;
  cellar_id_list.forEach(function (cellar) {
    loader.loadCellar(cellar, sku_list);
  });
  return this;
};

Loader.prototype.loadCellar = function (cellarid, sku_list) {
  var i, j, temparray;
  for (i = 0, j = sku_list.length; i < j; i += this.chunk_size) {
    temparray = sku_list.slice(i, i + this.chunk_size);
    // do whatever
    this.performRequest(cellarid, temparray);
  }
};

Loader.prototype.performRequest = function (cellarid, skus_chunk) {
  var loader = this;
  if (loader.instances > loader.max_instances) {
    setTimeout(function () {
      loader.performRequest(cellarid, skus_chunk);
    }, 100);
    return
  }
  this.instances += 1;
  fetch("https://apibodegas.loadingplay.com/v1/cellar/" + cellarid + "/product/?sku_list=" + skus_chunk.join(","),
  {
    headers: {
      "Authorization": "Bearer " + this.access_token
    }
  })
  .then((response) => response.json())
  .then((response) => {
    loader.instances -= 1;
    loader.callback(cellarid, response.products);
  });
};


Loader.prototype.done = function (callback) {
  this.callback = callback
};

export function StockLoader(access_token) {
  return new Loader(access_token);
}
