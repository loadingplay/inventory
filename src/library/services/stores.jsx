export class StoreLoader {
  constructor(access_token) {
    this.access_token = access_token;
  }

  loadCellars = async () => {
    let response,
      json_data;

    response = await fetch("https://apibodegas.loadingplay.com/v1/cellar", {
      headers: {
        "Authorization": `Bearer ${this.access_token}`
      }
    });
    json_data = await response.json();
    return json_data;
  }

  loadProducts = async (page, cellar_id, sku_filter) => {
    let response, json_data;

    response = await fetch(
      `https://replenishments.loadingplay.com/replenishment?items=100&page=${page}&cellar_id=${cellar_id}&search_word=${sku_filter}`,
      {
        "headers": {
          "Authorization": `Bearer ${this.access_token}`
        }
      });

    json_data = await response.json();
    return json_data;
  }

  setSelectedCellar = (cellar) => {
    localStorage.setItem("selected-cellar", JSON.stringify(cellar));
  }

  getSelectedCellar = () => {
    return JSON.parse(localStorage.getItem("selected-cellar"));
  }
}
