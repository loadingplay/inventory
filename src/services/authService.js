export default  {
    isLoggedIn(){
        let expiresAt = JSON.parse(window.localStorage.getItem('expiresAt'))
        if(!window.localStorage.getItem('token') || new Date().getTime() < expiresAt){
            return false;
        } else {
            return true;
        }
    },

    getCurrentToken(){
        let token = window.localStorage.getItem('token');
        return token;
    },

    logout() {
        window.localStorage.removeItem('token')
        window.localStorage.removeItem('expiresAt')
        window.location = 'https://accounts.loadingplay.com/auth/logout';
    }
  } 