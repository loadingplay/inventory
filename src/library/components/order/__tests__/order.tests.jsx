jest.mock("../../../services/orders");

import React from "react";
import { GenerateOrderButton } from "..";
import { shallow } from "enzyme";
import { PickerStore } from "../../../services";
import { Orders } from "../../../services/orders";


describe("GenerateOrderButton", () => {

  beforeEach(() => {
    PickerStore.clear(2);
    jest.clearAllMocks();
  });

  test("it should render the button from status", async () => {
    const wrapper = shallow(<GenerateOrderButton />);

    expect(wrapper).toMatchSnapshot("default");
    expect(wrapper.childAt(0).prop("disabled")).toBeFalsy();

    wrapper.setState({
      button_status: GenerateOrderButton.Statusses.GENERATING
    });
    expect(wrapper.childAt(0).prop('disabled')).toBeTruthy();
    expect(wrapper).toMatchSnapshot("loading");

    wrapper.setState({
      button_status: GenerateOrderButton.Statusses.DONE
    });
    expect(wrapper.childAt(0).prop('disabled')).toBeTruthy();
    expect(wrapper).toMatchSnapshot("done");
  });

  test("it should generate order data", () => {
    const wrapper = shallow(<GenerateOrderButton selectedCellar={2} />);

    expect(
      wrapper.instance()._generateOrderData()
    ).toMatchSnapshot("order-data-empty");

    PickerStore.set(2, "test-sku", 1, 10, 10, 10);
    expect(
      wrapper.instance()._generateOrderData()
    ).toMatchSnapshot("order-data-with-one-product");
  });

  test("it should send shipped order to api", async () => {
    const wrapper = shallow(<GenerateOrderButton selectedCellar={2} />);

    PickerStore.clear(2);
    Orders.prototype.create = jest.fn().mockReturnValue({});
    Orders.prototype.update = jest.fn().mockReturnValue({});

    await wrapper.instance()._sendShippedOrder();
    expect(Orders.prototype.create).toHaveBeenCalledWith({
      "extra_info": "{}",
      "origin": "replenishments",
      "payment_type": "guia",
      "products": "[]"
    });
    expect(Orders.prototype.update).toHaveBeenCalledTimes(0);

    // add products
    Orders.prototype.create = jest.fn().mockReturnValue({
      "status": "success",
      "order": {
        "id": 1
      }
    });
    PickerStore.set(2, "test-sku", 1, 10, 10, 10);
    await wrapper.instance()._sendShippedOrder();
    expect(Orders.prototype.create).toHaveBeenCalledWith({
      "extra_info": "{}",
      "origin": "replenishments",
      "payment_type": "guia",
      "products": "[{\"sku\":\"test-sku\",\"price\":0,\"name\":\"\",\"combination\":\"\",\"quantity\":1}]"
    });
    expect(Orders.prototype.update).toHaveBeenCalledWith(
      1, {"status": "despachado"}
    );
  });

  test("it should execute all sync within click", async () => {
    const wrapper = shallow(
      <GenerateOrderButton
        onOrderGenerated={jest.fn().mockReturnValue(1)}
      />
    );

    wrapper.instance()._sendShippedOrder = jest.fn().mockResolvedValue(1);

    let promise = wrapper.instance().handleClick();
    expect(wrapper.instance().state.button_status).toEqual(GenerateOrderButton.Statusses.GENERATING);
    expect(wrapper.instance()._sendShippedOrder).toHaveBeenCalledTimes(1);

    await promise;
    expect(wrapper.instance().props.onOrderGenerated).toHaveBeenCalledTimes(1);
    expect(wrapper.instance().state.button_status).toEqual(GenerateOrderButton.Statusses.DONE);
  });
});
