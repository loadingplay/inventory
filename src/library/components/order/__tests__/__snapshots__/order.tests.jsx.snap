// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerateOrderButton it should generate order data: order-data-empty 1`] = `
Object {
  "origin": "replenishments",
  "payment_type": "guia",
  "products": "[]",
  "reference_code": "uuidv4",
}
`;

exports[`GenerateOrderButton it should generate order data: order-data-with-one-product 1`] = `
Object {
  "origin": "replenishments",
  "payment_type": "guia",
  "products": "[{\\"sku\\":\\"test-sku\\",\\"price\\":0,\\"name\\":\\"\\",\\"combination\\":\\"\\",\\"quantity\\":1}]",
  "reference_code": "uuidv4",
}
`;

exports[`GenerateOrderButton it should render the button from status: default 1`] = `
<Fragment>
  <button
    className="btn btn-primary ml-1"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Generar guía
  </button>
  <OrderModal
    isOpen={false}
    onClose={[Function]}
    orderId={-1}
  />
</Fragment>
`;

exports[`GenerateOrderButton it should render the button from status: done 1`] = `
<Fragment>
  <button
    className="btn btn-primary ml-1"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Guía enviada
  </button>
  <OrderModal
    isOpen={false}
    onClose={[Function]}
    orderId={-1}
  />
</Fragment>
`;

exports[`GenerateOrderButton it should render the button from status: loading 1`] = `
<Fragment>
  <button
    className="btn btn-primary ml-1"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Generando...
  </button>
  <OrderModal
    isOpen={false}
    onClose={[Function]}
    orderId={-1}
  />
</Fragment>
`;
